src/Authenticator/CookieAuthenticator.php

```diff
@@ -0,0 +1,213 @@
+        ],
+        'cookie' => [
+            'name' => 'CookieAuth',
+            'expire' => null,
+            'path' => '/',
+            'domain' => '',
+            'secure' => false,
+            'httpOnly' => false
+        ],
+        'passwordHasher' => 'Authentication.Default',
+        'loginUrl' => null,
+        'useRegex' => false,
+        'checkFullUrl' => false
+    ];
+
+    /**
+     * {@inheritDoc}
+     */
+    public function __construct(IdentifierCollection $identifiers, array $config = [])
+    {
+        $this->_checkCakeVersion();
+
+        parent::__construct($identifiers, $config);
+    }
+
+    /**
+     * Checks the CakePHP Version by looking for the cookie implementation
+     *
+     * @return void
+     */
+    protected function _checkCakeVersion()
+    {
+        if (!class_exists('Cake\Http\Cookie\Cookie')) {
+            throw new RuntimeException('Install CakePHP version >=3.5.0 to use the `CookieAuthenticator`.');
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function authenticate(ServerRequestInterface $request, ResponseInterface $response)
+    {
+        $cookies = $request->getCookieParams();
+        $cookieName = $this->getConfig('cookie.name');
+        if (!isset($cookies[$cookieName])) {
+            return new Result(null, Result::FAILURE_CREDENTIALS_NOT_FOUND, [
+                'Login credentials not found'
+            ]);
+        }
+
+        $token = json_decode($cookies[$cookieName], true);
+
+        if ($token === null || count($token) !== 2) {
+            return new Result(null, Result::FAILURE_CREDENTIAL_INVALID, [
+                'Cookie token is invalid.'
+            ]);
+        }
+
+        list($username, $tokenHash) = $token;
+
+        $credentials = [
+            'username' => $username
+        ];
+        $identity = $this->identifiers()->identify($credentials);
+
+        if (empty($identity)) {
+            return new Result(null, Result::FAILURE_IDENTITY_NOT_FOUND, $this->identifiers()->getErrors());
+        }
+
+        if (!$this->_checkToken($identity, $tokenHash)) {
+            return new Result(null, Result::FAILURE_CREDENTIAL_INVALID, [
+                'Cookie token does not match'
+            ]);
+        }
+
+        return new Result($identity, Result::SUCCESS);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function persistIdentity(ServerRequestInterface $request, ResponseInterface $response, $identity)
+    {
+        $field = $this->getConfig('rememberMeField');
+        $bodyData = $request->getParsedBody();
+
+        if (!$this->_checkLoginUrl($request) || !is_array($bodyData) || empty($bodyData[$field])) {
+            return [
+                'request' => $request,
+                'response' => $response
+            ];
+        }
+
+        $data = $this->getConfig('cookie');
+        $value = $this->_createToken($identity);
+
+        $cookie = new \Cake\Http\Cookie\Cookie(
```

Comment:
Those FQCN should be moved to use statement ideally. Inline is usually more a code smell as it hides those things.

Reply:
I wasn't sure how should we handle the class names of "optional" packages.

Reply:
Hmm Symfony uses the `use` statements for potentially missing classes as well.
https://github.com/symfony/dom-crawler/blob/master/Crawler.php#L698

Reply:
Those can also be there IMO, if not found those will be ignored :)

Reply:
I wasn't aware that they'll be ignored and to lazy to test it. ☺️

Reply:
Fixed.